src/Microsoft.AspNetCore.Routing/Matching/DfaMatcher.cs

```diff
@@ -226,11 +258,100 @@ public sealed override Task MatchAsync(HttpContext httpContext, EndpointFeature
                 var constraint = constraints[i];
                 if (!constraint.Value.Match(httpContext, NullRouter.Instance, constraint.Key, values, RouteDirection.IncomingRequest))
                 {
+                    Logger.CandidateRejectedByConstraint(_logger, httpContext.Request.Path, endpoint, constraint.Key, values[constraint.Key]); 
                     return false;
                 }
             }
 
             return true;
         }
+
+        internal static class EventIds
+        {
+            public static readonly EventId CandidatesNotFound = new EventId(1000, "CandidatesNotFound");
+            public static readonly EventId CandidatesFound = new EventId(1001, "CandidatesFound");
+
+            public static readonly EventId CandidateRejectedByComplexSegment = new EventId(1002, "CandidateRejectedByComplexSegment");
+            public static readonly EventId CandidateRejectedByConstraint = new EventId(1003, "CandidateRejectedByConstraint");
+
+            public static readonly EventId CandidateNotValid = new EventId(1004, "CandiateNotValid");
+            public static readonly EventId CandidateValid = new EventId(1005, "CandiateValid");
+        }
+
+        private static class Logger
+        {
+            private static readonly Action<ILogger, string, Exception> _candidatesNotFound = LoggerMessage.Define<string>(
+                LogLevel.Debug,
+                EventIds.CandidatesNotFound,
+                "No candidates found for the request path '{Path}'");
```

Comment:
Log messages elsewhere end with a full stop. Repeat for all log messages.

Reply:
That's not true, I haven't heard such a guideline in the past no followed it elsewhere

Reply:
I'm removing the periods from the other logging in this repo for consistency. The most visible and oldest parts of our stack (logging that happens for all projects at info level, like Hosting, MVC) don't use periods. 

If you think this is the wrong decision then that's a topic for the broader engineering team. We haven't had a defined practice in this area.

Reply:
ok!